### ISSUE TITLE
How to reproduce Dutch relation classification results

### ISSUE THREAD

**Issue by erzaliator** [gucorpling/DisCoDisCo#2]
The paper achieves an accuracy of **55.21 and 52.15** (with and without) for nld.rst.nldt relation classification.

I was able to get the following results by seeding for different values 
AllenNLP parameters modified in rel/flairclone.jsonnet:

- random_seed : as in table
- numpy_seed : as in table
- pytorch_seed : as in table
- trainer patience: 99

| With features  | Without features | random_seed | numpy_seed | pytorch_seed |
| ------------- |:-------------:|:-------------:|:-------------:| ------------|
| 0.5061      | 0.5460     | 13370 | 1337 | 133 |
| 0.5215      | 0.5583     | 23 | 23 | 23 |
| 0.5460      | 0.5368     | 50 | 50 | 50 |             
| 0.5491      | 0.5613     | 123 | 123 | 123 |

This brings the 4-run average to be  **0.530675 and 0.5506** (with and without) on the Dutch relation classification task. 

How can I reproduce the paper results?

**Comment by lgessler** [comment]
Hi @erzaliator! It seems like you're doing everything right--our relation classification model is just disappointingly high-variance. During our development process we noticed different results across machines (pointing to hardware differences as a possible source of training instability) and even on a single machine (cf. [PyTorch docs](https://pytorch.org/docs/stable/notes/randomness.html)), leading to differences of as much as 5% F1 between runs.

Looking at our results table, it looks like for nldt.rst.nldt in particular we were able to consistently achieve 55.21 and 52.15 (respectively) on all 5 runs. It's impossible to know why your scores are different, but one possible reason could be differences in underlyingly nondeterministic algorithms used in PyTorch operations. If you have a different GPU available, you could try running on that, or you could also try setting PyTorch's [`use_deterministic_algorithms`](https://pytorch.org/docs/stable/notes/randomness.html) flag.

Pinging @janetlauyeung in case she has more to say, since I believe she was the one to get the final results for nld.

**Comment by janetlauyeung** [comment]
Hi @erzaliator - In addition to what @lgessler mentioned, which I agreed regarding the possible factors that lead to training instability, I just double checked our own record of the final results for the nld.rst.nldt dataset, the scores are reported correctly. This is unlikely, but could you please double check that the model that you referred to as "with features" is in fact "with features" and vice versa? It looks that the score tendency is consistently flipped (except for the third run).

**Comment by erzaliator** [comment]
@janetlauyeung I have double checked the code and I am correctly reporting the results.

@lgessler  I am running the code on a single core of Tesla V100.  

@both, I also modified the code to the original patience = 12 (which was previously 99) to get the following results

| With features | Without features | random_seed | numpy_seed | pytorch_seed |
| ----------------- | -------------------- | ---------------- | --------------- | ----------------- |
| 0.5061            | 0.546                | 13370            | 1337            | 133               |
| 0.4969            | 0.5798               | 23               | 23              | 23                |
| 0.5583            | 0.5307               | 50               | 50              | 50                |
| 0.5215            | 0.5491               | 123              | 123             | 123               |

This brings the 4-run average to be 0.5207 and 0.5514 (with and without) again getting lower results while using the features.

These are my config files: [configs/rel/flair_clone.jsonnet ](https://github.com/erzaliator/RadomDump/blob/main/with_features_flair_clone.jsonnet) (with features) and [configs/rel/flair_clone.jsonnet ](https://pages.github.com/) (without features). Note that I have added the seed parameters and emptied the feature tuple for nld. I hope they are correct.

Could you provide the config file for your runs? Did you change the seed every run?

**Comment by lgessler** [comment]
Those config files are the ones we used for our runs. We did not change our seeds in between runs--we noticed enough variance even without the seeds, and we were only trying to control for that. So whatever AllenNLP's default seeds are are the ones you should use.

**Comment by erzaliator** [comment]
@lgessler  Instead of running the code on a Tesla V100, I ran the default AllenNLP seeds on a single core GTX TITAN X to get 0.5429 and 0.5184 with and without features respectively. 

 Notably, I got the same results for each of the four runs (even the class-wise breakdown was identical).Thanks!
